### ISSUE TITLE
process_two_files_with_parallel_output() function needed

### ISSUE THREAD

**Issue by wjlei1990** [SeismicData/pyasdf#21]
Hi Lion,

I think we have reached the point when we need the parallel output for adjoint source for the preprocessing workflow.

Is it very difficult to write such a function for adjoint source writer? I am thinking about looking into this issue this weekend? Are you comfortable with me having some attempts?

**Comment by wjlei1990** [comment]
Actually, I think, writing the adjoint source is quite a specific task. People can actually generate all kinds of different data when using the __process_two_files_with_parallel_output__. How do you think this should be dealt with? Do we need to develop different writer for different cases? Or it might be able to fit in one common function?

**Comment by krischer** [comment]
Hi Wenjie,

it is not super difficult but tedious as one has to be extremely careful. A generic version is definitely possible and I envision something like this (we might have to think about this a bit more):

```python
def process(...):
    ...
    # Can return a stream object in which case it will write to the /Waveform group.
    return {"object": st, "type": "waveforms", tag="processed"}
    # Or an auxiliary data piece.
    return {"object": data_array, "type": "auxiliary_data",
            "path": "/AdjointSources/Event_a/Station_E_D",
            "parameters": {...}}
```

Then the API can decide what it wants to actually write to disc. In terms of implementing it: you can lean on the `_dispatch_processing_mpi()` method: https://github.com/SeismicData/pyasdf/blob/22d65ad18933e9d12a3ad1e0641e5a81a6ebd90b/pyasdf/asdf_data_set.py#L1574

I think it should be possible to come up with some generic parallel I/O function that works for all the various kinds of processing functions we have, e.g. `process()`, `process_two_files()`, `process_each_with_each()`. All of them will then gain parallel I/O capabilities and will just be specializations of the generic function.

I can help with it around the end of next week. I'll do it in exchange for the paper chapter ;-P


If you want to give it a try:

1. Rip out the `process()` method from the `ASDFDataSet` object and place it in a separate file like done in #20.
2. Generalize that function to work with any number of input files.
3. Add the ability to also write auxiliary data in parallel.
4. Profit :-)


The main difficulty is that each rank must know what every other rank wants to write before it can actually write to the file. The `process()` method currently works like this:

1. The master node (rank 0) collects a list of all jobs and distributes those to all workers that want some work.
2. Workers will request new jobs as long as their buffer is not full. Each worker buffers the results until a certain amount of memory.
3. Once a certain number of workers have filled their buffers, the master node triggers a metadata collection, e.g. each rank tells each other the size, dtypes, groups, and attributes to be written.
4. Each rank writes that metainformation from every rank.
5. Then each rank can independently write the data.
6. Rinse and repeat until no more jobs are left.

The hard part is making sure that no deadlocks exist.

**Comment by wjlei1990** [comment]
Deal! Deal!

:)

**Comment by wjlei1990** [comment]
is `add_auxiliary_data` works in mpi?

Seems not...

I am trying to use `add_auxiliary_data` to add adjoint sources to asdf file, but it hangs at this function.

**Comment by krischer** [comment]
Yea you cannot call this with MPI...you need to split it up like `add_waveforms()` is internally split up.

**Comment by wjlei1990** [comment]
Got it done.

I will try to clean up the code and let you review it by creating a pull request.

**Comment by wjlei1990** [comment]
I will close this issue because I reopened another thread here:
https://github.com/SeismicData/pyasdf/issues/31
